fastem calibration 2: dark/gain calibration per scintillator

Sabrina85 opened the pull request:
* add dark/gain to calibration manager
* make controller independent of each other
* make calibration controller using a prefix argument to be able to use same class for all calibration steps
* make the gauge having a fixed size in the GUI
* use short version for text label showing the progress
* fix: ROC size should be calculated based on multibeam resolution and pixelsize
* allow adding/removing ROCs
* synchronize gui with main model (active scintillators and calibration regions defined)
* remove unnecessary properties in tabs

tepals commented on src/odemis/gui/cont/fastem_acq.py:
```diff
@@ -581,32 +563,231 @@ def _on_alignment_done(self, future):
+        for roc_num in sorted(self._tab_data.calibration_regions.value.keys()):
+            # check if calibration region (ROC) on scintillator is set
+            roc = self._tab_data.calibration_regions.value[roc_num]
+            if roc.coordinates.value != stream.UNDEFINED_ROI:
```
The comment on line 668 is not completely clear to me with respect line. Can you clarify that comment or add another comment here, explaining what `roc.coordinates.value != stream.UNDEFINED_ROI` means. Does it mean the ROC is not set?

Sabrina85 replied:
`roc.coordinates.value != stream.UNDEFINED_ROI` means that the ROC is set, if it is undefined, it is not set.

tepals commented on src/odemis/gui/cont/fastem_acq.py:
```diff
@@ -581,32 +563,231 @@ def _on_alignment_done(self, future):
+                f.add_done_callback(partial(self._on_calibration_done, roc_num=roc_num))
+                futures[f] = t
+
+        # if futures:
```
what happens if futures is an empty dict?

Sabrina85 replied:
The gui is blocked. I am just wondering if the dict can ever be empty.

Sabrina85 replied:
@pieleric In which case would the dict be empty?

tepals commented on src/odemis/gui/cont/fastem_acq.py:
```diff
@@ -581,32 +563,231 @@ def _on_alignment_done(self, future):
+        #     self._main_data_model.is_acquiring.value = False
+        #     self._reset_calibration_gui("Calibrations failed (see log panel).", level=logging.WARNING)
+
+    # @call_in_wx_main
```
why is this commented?

tepals commented on src/odemis/gui/cont/fastem_acq.py:
```diff
@@ -581,32 +563,231 @@ def _on_alignment_done(self, future):
+            self.is_aligned.value = False  # don't enable ROA acquisition
+            logging.exception("Calibration failed with %s.", ex)
+            self._update_calibration_controls("Calibration failed")
+            # return
```
I think all these `# return`'s can be removed right?

Sabrina85 replied:
I am not sure. 
The cancelling is not raised correctly for some reason atm. 
And in the ROA acquisition and overview acquisition there is a return. 
@pieleric can you comment please?

Sabrina85 replied:
What happens at the moment is that when I cancel the calibration, it raises the CancellationError in the callback of the single future, but then in the batch future it is not receiving the CancellationError.

tepals commented on src/odemis/gui/cont/fastem_acq.py:
```diff
@@ -581,32 +563,231 @@ def _on_alignment_done(self, future):
         else:
-            duration = align.fastem.estimate_calibration_time(self.calibrations)
-            self.gauge_label.SetLabel(str(duration) + " seconds")
+            # check if at least one roc was selected
+            if all(roc.coordinates.value == stream.UNDEFINED_ROI
+                   for roc in self._tab_data.calibration_regions.value.values()):
```
The comment does not match the actual code. By reading the comment I would think `all(roc.coordinates.value == stream.UNDEFINED_ROI for roc in self._tab_data.calibration_regions.value.values()) = True` would mean that at least one roc was selected, but if I read the code it means that no roc was selected. You could update it to:
```
            # check if at least one roc was selected
            if any(roc.coordinates.value != stream.UNDEFINED_ROI
                   for roc in self._tab_data.calibration_regions.value.values()):
                self._update_calibration_controls()
            else:
                self._update_calibration_controls("No calibration region selected.", False)
```

tepals commented on src/odemis/gui/cont/project.py:
```diff
@@ -336,100 +349,174 @@ def _on_text(self, evt):
     def fit_view_to_bbox(self):
         """
-        Zoom in to calibration region. Calibration region is in the center and takes up 1/100 of viewport area.
+        Zoom into calibration region. Calibration region is centered in the scintillator with the
```
I think actually "Zoom in on calibration region" would fit better.

tepals commented on src/odemis/gui/cont/project.py:
```diff
@@ -336,100 +349,174 @@ def _on_text(self, evt):
+        corresponding number and takes up approximately 1/100 of the viewport area.
         """
-        logging.debug("Zooming in to calibration region %s.", self.model.name.value)
+        logging.debug("Zooming into calibration region %s.", self.calib_model.name.value)
```
I think actually "Zoom in on calibration region" would fit better.

tepals commented on src/odemis/gui/cont/project.py:
```diff
@@ -336,100 +349,174 @@ def _on_text(self, evt):
+        l, t, r, b = self.calib_model.coordinates.value
+        size = (self._tab_data.main.multibeam.resolution.value[0] * self._tab_data.main.multibeam.pixelSize.value[0],
+                self._tab_data.main.multibeam.resolution.value[1] * self._tab_data.main.multibeam.pixelSize.value[1])
+        cnvs.fit_to_bbox([l - 5 * size[0], t - 5 * size[1], r + 5 * size[0], b + 5 * size[1]])
```
Why do you do `5*size`?

Sabrina85 replied:
It defines how much we zoom in on the ROC - so that there is still some space shown around the ROC. I added a comment.

tepals commented on src/odemis/gui/cont/project.py:
```diff
@@ -336,100 +349,174 @@ def _on_text(self, evt):
-        btn.SetForegroundColour(wx.GREEN)
+        """
+        Called when one of the 9 single region of calibration (ROC) buttons is triggered.
+        A ROC can be added or removed by clicking again. If a ROC is added, an overlay will be
```
Maybe a typo: I pronounce ROC as "arr ooo ceee" so I would write "an ROC", if you pronounce it as "rock" you can leave it as is.

https://www.quora.com/When-using-an-abbreviation-starting-with-the-letter-R-is-it-a-or-an-For-example-I-played-an-RTG-game#:~:text=It%20depends%20whether%20you%20pronounce,correct%20to%20use%20a%20instead.

tepals commented on src/odemis/gui/cont/tabs.py:
```diff
@@ -1809,13 +1795,31 @@ def __init__(self, name, button, panel, main_frame, main_data):
         )
 
-        # Controller for acquistion settings panel
+        # FIXME instantiate FastEMCalibrationRegionsController in FastEMScintillatorCalibrationController inherit from?
```
I do not understand this sentence, can you improve it?

tepals reviewed (changes requested):
I also pulled the code and the GUI looks good. I noticed that the calibration 2 button is enable, even when calibration 1 has not been run. Is that on purpose?

Sabrina85 commented:
> 

yes that does not work yet. For this I need to make a fancier implementation for when the is_aligned. I will add a TODO somewhere.

Sabrina85 commented on src/odemis/acq/align/fastem.py:
```diff
@@ -223,6 +252,8 @@ def run(self):
             logging.debug("Calibrations finished.")
 
+        return self.asm_config
+
     def run_calibrations(self, calibration):
```
yes, but as I am already working on the field corrections I will remove it already now.

Sabrina85 commented on src/odemis/gui/comp/miccanvas.py:
```diff
@@ -1855,6 +1855,14 @@ def add_calibration_overlay(self, coordinates, label, colour=gui.FG_COLOUR_WARNI
         overlay.active.value = False  # no need to activate by default
         return overlay
 
+    def remove_calibration_overlay(self, overlay):
+        """
+        overlay (FastEMROAOverlay): overlay to be deleted
+        """
+        overlay.active.value = False  # deactivating the overlay avoids weird behaviour
+        self.remove_world_overlay(overlay)
+        wx.CallAfter(self.request_drawing_update)
```
I think you are right that we can reuse it. I just tested it and it seems to work fine (there are not dump questions ;) ).

Sabrina85 commented on src/odemis/gui/cont/fastem_acq.py:
```diff
@@ -581,32 +563,231 @@ def _on_alignment_done(self, future):
+
+                # calculate the center position of the ROC (half field right/bottom
+                # compared to top/left corner in view)
+                l, t, _, _ = roc.coordinates.value  # (l, t, r, b)
```
Is top = ymin?

tepals replied:
For the ROA the coordinates we get are in the sample carrier system, which corresponds to the component with role='stage'. There `xmin, ymin, xmax, ymax` is equal to `l, t, r, b`

Sabrina85 commented on src/odemis/gui/cont/project.py:
```diff
@@ -336,100 +350,175 @@ def _on_text(self, evt):
+        # update ROC buttons
+        self._on_coordinates()
+
+    def _update(self):
+        pass
```
Oh that should go. This was a left over from discussions.

pieleric commented on src/odemis/gui/cont/fastem_acq.py:
```diff
@@ -431,6 +430,7 @@ def _reset_acquisition_gui(self, text=None, level=None):
         else:
             self.update_acquisition_time()
 
+    @call_in_wx_main  # call in main thread as changes in GUI are triggered
```
No, `on_acquisition()` is called when the "acquire" button is pressed. (see the "evt" argument). So it's already called in the main GUI thread. It shouldn't be called delayed.

pieleric commented on src/odemis/gui/cont/project.py:
```diff
@@ -291,6 +297,7 @@ def __init__(self, name, roc, colour, tab_data, project_panel, view_ctrl):
         self.overlay = canvas.add_roa_overlay(self.model.coordinates, colour)
         self.overlay.active.subscribe(self._on_overlay_active)
 
+    # FIXME how taken care of that his is run in the main gui thread?
```
Add a note in the docstring "Should be called from the main GUI thread."  (and then make sure that the callers are running in the main GUI thread.

pieleric commented on src/odemis/gui/cont/project.py:
```diff
@@ -215,7 +221,7 @@ def _add_roa_ctrl(self, coords):
         self._project_bar.enable_buttons(True)
 
-    # Should be called from GUI main thread
+    # already running in main GUI thread as it receives event from GUI
```
Well, it's not directly receiving events from the GUI. Looking at the code, I see it's called from `_on_btn_remove()`, which is receiving a GUI event. I'd suggest to add a docstring saying "Must be called from within the main GUI thread".

pieleric commented on src/odemis/gui/cont/project.py:
```diff
@@ -368,6 +379,7 @@ def __init__(self, number, tab_data, view_ctrl):
 
         self.overlay = None
 
+    @call_in_wx_main  # call in main thread as changes in GUI are triggered
```
Indeed, now I see that `canvas.fit_to_bbox()` does use some GUI stuff (while it also tries to not depend on GUI stuff!). So we could add a note in the `fit_to_bbox()` docstring that this function should be called within the main GUI thread.
An alternative to adding the decorator here would be to call `wx.CallAfter(cnvs.fit_to_bbox, [...])` (as it's the last function called, and we don't need the result)

Sabrina85 replied:
I tried the wx.CallAfter (and no decorator) but then I get this error:
`11:37:42 ERROR   main           : Traceback (most recent call last):
  File "/home/sabrina/development/odemis/src/odemis/gui/cont/project.py", line 485, in _on_button
    self.roc_ctrls[num].fit_view_to_bbox()  # Zoom to calibration region
  File "/home/sabrina/development/odemis/src/odemis/gui/cont/project.py", line 399, in fit_view_to_bbox
    wx.CallAfter(cnvs.fit_to_bbox([xmin - 5 * size[0], ymin - 5 * size[1], xmax + 5 * size[0], ymax + 5 * size[1]]))
  File "/usr/lib/python3/dist-packages/wx/core.py", line 3183, in CallAfter
    assert callable(callableObj), "callableObj is not callable"
AssertionError: callableObj is not callable`

pieleric replied:
Please note the syntax. `CallAfter()` takes the function as first argument, and the arguments as following arguments. The way you wrote it immediately called `fit_to_bbox()`, then took the return value (None), and passed it to `CallAfter()`.
You need to write something like `wx.CallAfter(cnvs.fit_to_bbox, [...])`.

Sabrina85 commented:
- replaced the viewport with the view only
- added test case for calibration manager: dark/gain calibration
- fixed/updated the wx calls in main thread
There is still an issue when you really quickly add and remove projects/ROAs it sometimes fails. I think it has to do with the update time of 1s.

tepals commented on src/odemis/gui/comp/miccanvas.py:
```diff
@@ -656,6 +656,8 @@ def fit_to_bbox(self, bbox):
         bbox (4 floats): bounding box to be shown, defined as minx, miny, maxx,
           maxy positions in m.
+
+        Note: Should be called from main GUI thread.
```
If it is already called from the main GUI thread I would update this comment to: `Note: Is called from main GUI thread`. If it should be forced to be called from the main GUI thread, I think you need to add the `@call_in_wx_main` decorator to this function.

Sabrina85 replied:
I added this comment in concordance with Éric's review. The "should" indicates that every caller should already be running in the main GUI thread, so also potential future callers. I think that is a bit the difference to "Is" - this does not highlight the need that I need to make sure the caller is already running in the main GUI thread. 
I can extend the comment and add "Caller should run in main GUI thread".

I don't know if the decorator makes sense here. The part that changes the GUI is the drawing update, which is in a wx.CallAfter call. From what I understand this makes sure that this call is only executed after the current and pending event handlers have been completed. But I don't understand why we still need to make sure that the caller of this method runs in the main gui thread. @pieleric can you clarify?

tepals replied:
Then can you maybe add what call makes sure that this is called from the main GUI thread?

Sabrina85 commented on src/odemis/gui/cont/project.py:
```diff
@@ -336,100 +350,176 @@ def _on_text(self, evt):
+        Called when one of the 9 single region of calibration (ROC) buttons is triggered.
+        An ROC can be added or removed by clicking again. If a ROC is added, an overlay will be
+        created in the center of the corresponding viewport. Size of the ROC is one single field.
+        If a ROC is removed, the coordinates are reset to undefined.
```
Yes indeed, this behavior we want in future.

Sabrina85 commented on src/odemis/gui/cont/project.py:
```diff
@@ -336,100 +350,176 @@ def _on_text(self, evt):
         """
-        Zoom in to calibration region. Calibration region is in the center and takes up 1/100 of viewport area.
+        Zoom in on calibration region (ROC). Calibration region is centered in the scintillator with the
+        corresponding number and takes up approximately 1/100 of the viewport area.
```
I rephrased:
`Zoom in on calibration region (ROC). Calibration region has the size of a single field
        image, which is approximately 1/100 of the viewport area (= size of a single scintillator)
        and is centered in the scintillator with the corresponding number.`

Sabrina85 commented on src/odemis/acq/fastem_conf.py:
```diff
@@ -143,3 +143,22 @@ def configure_scanner(scanner, mode):
+    if "cellDarkOffset" in roc.value.parameters:
+        detector.cellDarkOffset.value = roc.value.parameters["cellDarkOffset"]
+    else:
+        logging.warning("Region of calibration doesn't have dark offset parameters.")
```
Not necessarily. It is two-fold. If it cannot find any parameters it indeed it will use the currently set values on the VAs. BUT, this method is called with two different ROCs. The first one will only have dark/gain parameters, so it will never have cell translation parameters and will always trigger the warning. However, it will not use the VA settings for cell translation parameters as the same function is also called with the ROC that actually contains the cell translation parameters. 
So I cannot add that the VA values are used. We can maybe change how the method is used by the 2 ROCs. But I don't have a good idea yet.

Sabrina85 replied:
Will change input arg to list of ROCs in pr for calibration step 3.
